src/operator/timeout.ts

```diff
@@ -65,18 +68,34 @@ class TimeoutSubscriber<T> extends Subscriber<T> {
   private static dispatchTimeout(state: any): void {
     const source = state.subscriber;
     const currentIndex = state.index;
-    if (!source.hasCompleted && source.previousIndex === currentIndex) {
+    if (source.previousIndex === currentIndex) {
       source.notifyTimeout();
     }
   }
 
   private scheduleTimeout(): void {
-    let currentIndex = this.index;
-    this.scheduler.schedule(TimeoutSubscriber.dispatchTimeout, this.waitFor, { subscriber: this, index: currentIndex });
+    const currentIndex = this.index;
+    const timeoutState = { subscriber: this, index: currentIndex };
+
+    this.cancelTimeout();
```

Comment:
So here it seems like we can do this:

```es6
const { action } = this;
if (!action) {
  this.add(this.action = this.scheduler.schedule(...));
} else {
  action.schedule(this.waitFor, timeoutState);
}
```

Instead of scheduling an action, throwing it away, and scheduling again, we can recycle the Action and reschedule it for a new due time. This should be both more efficient, and allow us to remove more code from this file.